### DISCUSSION TITLE
I can't get BasicHttpBinding with WSHttpBinding to work together.

### DISCUSSION THREAD

**Discussion by VirReal1** [CoreWCF/CoreWCF#1467] [Q&A] [answered]
Hello. I'm trying to migrate my WCF project from .NET Framework to .NET Core 8 using CoreWCF but I'm experiencing some issues.

I can't define BasicHttpBinding with WSHttpBinding or vice versa. Only one of them works and the client can't access the other endpoint.

```
    BasicHttpBinding noneBasicHttpBinding = new BasicHttpBinding(BasicHttpSecurityMode.None);

    BasicHttpBinding transportBasicHttpsBinding = new BasicHttpBinding(BasicHttpSecurityMode.Transport);
    transportBasicHttpsBinding.Security.Transport.ClientCredentialType = HttpClientCredentialType.None;

    WSHttpBinding noneWSHttpBinding = new WSHttpBinding(SecurityMode.None);

    WSHttpBinding transportWSHttpsBinding = new WSHttpBinding(SecurityMode.Transport);
    transportWSHttpsBinding.Security.Transport.ClientCredentialType = HttpClientCredentialType.None;

    serviceBuilder.AddService<BaseWrapperImpl>()
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneBasicHttpBinding, "/Service.svc")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportBasicHttpsBinding, "/Service.svc")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneWSHttpBinding, "/Service.svc/soap12")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportWSHttpsBinding, "/Service.svc/soap12");

    var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
    serviceMetadataBehavior.HttpGetEnabled = true;
    serviceMetadataBehavior.HttpsGetEnabled = true;
```

/Service.svc?wsdl includes both of the endpoints and wsdl looks correct but I can't send any request to /soap12 endpoint. Also If I remove BasicHttpBinding, WSHttpBinding works correctly. It seems like only one of the bindings works and the second just does not register. My previous Web.config is like this.

```
      <serviceActivations>
               <add relativeAddress="Service.svc" service="ApiWrapper.Implementations.BaseWrapperImpl"/>
      </serviceActivations>
	  <service name="ApiWrapper.Implementations.BaseWrapperImpl">
		<endpoint bindingConfiguration="defaultBasicHttpsBinding" address="" binding="basicHttpBinding" contract="ApiWrapper.Services.I_Authentication" />
		<endpoint bindingConfiguration="defaultBasicHttpBinding" address="" binding="basicHttpBinding" contract="ApiWrapper.Services.I_Authentication" />
		<endpoint bindingConfiguration="defaultwsHttpBinding" address="soap12" binding="wsHttpBinding" contract="ApiWrapper.Services.I_Authentication" />
		<endpoint bindingConfiguration="defaultwsHttpsBinding" address="soap12" binding="wsHttpBinding" contract="ApiWrapper.Services.I_Authentication" />
	</service>

        <binding name="defaultwsHttpBinding">
          <security mode="None"/>
        </binding>
        <binding name="defaultwsHttpsBinding">
          <security mode="Transport">
            <transport clientCredentialType="None"/>
          </security>
        </binding>
      </wsHttpBinding>
      <basicHttpBinding>
        <binding name="defaultBasicHttpsBinding">
          <security mode="Transport">
            <transport clientCredentialType="None"/>
          </security>
        </binding>
        <binding name="defaultBasicHttpBinding">
          <security mode="None"/>
        </binding>
```

I just dicovered that If I change the position of WSHttpBinding with BasicHttpBinding, both endpoints work but the Service metadata is now located at /Service.svc/soap12 endpoint.

```
    serviceBuilder.AddService<BaseWrapperImpl>()
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneWSHttpBinding, "/Service.svc/soap12")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportWSHttpsBinding, "/Service.svc/soap12")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneBasicHttpBinding, "/Service.svc")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportBasicHttpsBinding, "/Service.svc");
```

**Comment by VirReal1** [comment]
So I found a solution that is sufficient for the time being.

```
    serviceBuilder.AddService<BaseWrapperImpl>()
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneWSHttpBinding, "/Service.svc/soap12")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportWSHttpsBinding, "/Service.svc/soap12")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(noneBasicHttpBinding, "/Service.svc")
             .AddServiceEndpoint<BaseWrapperImpl, I_Authentication>(transportBasicHttpsBinding, "/Service.svc");
```

```
    var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
    serviceMetadataBehavior.HttpGetEnabled = true;
    serviceMetadataBehavior.HttpGetUrl = new Uri("/Service.svc", UriKind.Relative);
    serviceMetadataBehavior.HttpsGetEnabled = true;
    serviceMetadataBehavior.HttpsGetUrl = new Uri("/Service.svc", UriKind.Relative);
```

by manually configuring the ServiceMetadataBehaviour object, I have managed to mimic the web.config successfully but any answer about why changing the places between /Service.svc and /Service.svc/soap12 endpoints cause an issue will be much appreciated.

**Reply by mconnew** [reply, marked as answer]
Here's why you are seeing the behavior that you observed. The endpoints are mapped using a mechanism similar to app.Map, where it examines the path (and the scheme, http vs https) to see if the request matches. It's a simple prefix match because things like WebHttpBinding can require other parameters or even path components after the path you specify. This is a first match wins behavior, so if /Service.svc is defined first, it will also match for /Service.svc/soap12. This means you need to declare the endpoint with the more specific path first.  

The next problem you had was the path of the wsdl. In WCF, you are forced to declare a base path. So you would specify "http://localhost/Service.svc" and "https://localhost/Service.svc" as base paths to the ServiceHost constructor. If you are IIS hosted, this is done automatically as requesting the .svc file is the entry point to starting the service, so gets passed to WCF at initialization. When hosting in asp.net core, there are implicit base paths based on the listening addresses. Except in a few situations, the path part of the url would be the root, ie /. It's almost certain you don't want the wsdl to be hosted at the root of the site, so there's a heuristic where it searches for an endpoint with the correct listening scheme (e.g. http) and uses the first one to match. When you have to put you /soap12 path first for the earlier reason, this is the first one that matches.  

As an FYI is an overload of AddService which takes a delegate which allows you to set the base paths. You can do this:
```csharp
serviceBuilder.AddService<BaseWrapperImpl>(options =>
    {
        options.BaseAddresses.Clear();
        options.BaseAddresses.Add("http://localhost/Service.svc");
        options.BaseAddresses.Add("https://localhost/Service.svc");
    });
```
This allows you to use relative addresses on the AddServiceEndpoint paths. I don't think I did this, but it might be an idea that if the base address is explicitly set, then use that as the default wsdl uri.

**Reply by VirReal1** [reply]
Thanks for the reply.

I'm trying to host it using IIS and Docker and so far I don't have any issues except that the docker one causes some issues regarding the generated wsdl file. If I generate the wsdl file by using the https scheme, http scheme is the hosted address of the website (i.e. localhost:8080) and vice versa. This is pretty much expected, and I think I need to use a reverse proxy to fix this problem.  I don't want to set BaseAddress directly because I need to be able to deploy to many servers without changing any of the configs.

Thanks again. I'm marking this as resolved.
